### PR TITLE
[Security] add PasswordEncoderInterface::needsRehash()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Split from #31153, with tests.

### CODE REVIEW

**Review by Simperfit — APPROVED** [review]

**Review by chalasr — APPROVED** [review]
Could be worth an UPGRADE entry as not implementing them will break in 5.0

**Comment by nicolas-grekas** [comment]
friendly ping @symfony/deciders

File: src/Symfony/Component/Security/Core/Encoder/SodiumPasswordEncoder.php [hunk 1]
```diff
@@ -99,4 +99,20 @@ public function isPasswordValid($encoded, $raw, $salt)
 
         throw new LogicException('Libsodium is not available. You should either install the sodium extension, upgrade to PHP 7.2+ or use a different encoder.');
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function needsRehash(string $encoded): bool
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Encoder/SodiumPasswordEncoder.php]: Should we add a check for BCrypt passwords too?

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Encoder/SodiumPasswordEncoder.php]: It's deprecated so I don't think we should improve it further.

**Review by Tobion — APPROVED** [review]
Agree with this. It's best practice to have the hash contain the algo, salt, and options and then determine based on this if a rehash is needed.
But going this way contradics the old way of doing that using a salt parameter and `UserPasswordEncoderInterface`. So I think we should deprecate 
- the salt paramter in https://github.com/symfony/symfony/blob/0119d217cad9cac6b2b6461c62fd879b64f16406/src/Symfony/Component/Security/Core/Encoder/PasswordEncoderInterface.php#L34
- https://github.com/symfony/symfony/blob/0119d217cad9cac6b2b6461c62fd879b64f16406/src/Symfony/Component/Security/Core/User/UserInterface.php#L68
- maybe also https://github.com/symfony/symfony/blob/0119d217cad9cac6b2b6461c62fd879b64f16406/src/Symfony/Component/Security/Core/Encoder/UserPasswordEncoderInterface.php#L21

**Comment by chalasr** [comment]
Thank you @nicolas-grekas.

**Pull request merged**